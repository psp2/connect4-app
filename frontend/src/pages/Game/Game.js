import React, { useState } from "react";
import "./Game.scss";
import { Redirect } from "react-router-dom";
import Board from "../../components/Board/Board";
import PlayerIcon from "../../components/PlayerIcon/PlayerIcon.js";
import InvalidMove from "../../components/InvalidMove/InvalidMove.js";

function Game() {
  // TEMP; FOR DEMO DISPLAY
  const temp_6 = [
    [0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0],
  ];
  // TEMP; FOR DEMO DISPLAY
  const temp_7 = [
    [0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0],
  ];
  // TEMP; FOR DEMO DISPLAY
  const temp_8 = [
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
  ];

  // Must add API call here
  const [gameState, setGameState] = useState([1, temp_8]);
  // Used to determine if the last move was valid or not
  const [prevPlayer, setPrevPlayer] = useState(1);
  const [lastMoveValid, setLastMoveValid] = useState(true);
  const [endGame, setEndGame] = useState(false);

  if (endGame) {
    return <Redirect to="/gameover" />;
  }

  function updateGameState(col) {
    // TODO: Make API Call and set game state accordingly!

    // TEMP; FOR WEEK 1 DEMO ONLY
    const api_res = [
      1,
      [
        [0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 2, 0, 1, 0, 0, 0],
        [0, 0, 2, 1, 1, 0, 0, 0],
        [0, 2, 1, 2, 2, 1, 0, 0],
      ],
    ];
    setGameState(api_res);
    // If the returned board is empty, the game is over.
    if (api_res[1] === []) {
      setEndGame(true);
    }
    // If the player hasn't changed between moves, the move was invalid.
    if (api_res[0] === prevPlayer) {
      setLastMoveValid(false);
    } else {
      setPrevPlayer(api_res[0]);
      setLastMoveValid(true);
    }
  }

  return (
    <div className="Game">
      <div className="GameZone">
        <div className="Player1_Zone">
          <PlayerIcon
            playerNumber={1}
            playerName="Prashant"
            endGame={setEndGame}
          />
        </div>
        <div>
          <Board
            player={gameState[0]}
            gameState={gameState[1]}
            makeMove={updateGameState}
          />
        </div>
        <div className="Player2_Zone">
          <PlayerIcon
            playerNumber={2}
            playerName="Vivek"
            endGame={setEndGame}
          />
        </div>
      </div>
      <div className="CurrentPlayer">Player Turn: {gameState[0]}</div>
      {!lastMoveValid && (
        <InvalidMove invalidMessageReceived={setLastMoveValid} />
      )}
    </div>
  );
}

export default Game;
