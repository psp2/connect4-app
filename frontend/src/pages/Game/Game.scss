.Game {
  height: 100vh;
  width: 100vw;
  margin: auto;
  background-color: black;
  display: flex;
  flex-direction: column;
}

.GameZone {
  display: flex;
  flex-direction: row;
}

.Player1_Zone {
  position: absolute;
  top: 5vh;
  left: 2.5vw;
  display: flex;
  flex-direction: column;
}

.Player2_Zone {
  position: absolute;
  top: 5vh;
  right: 2.5vw;
  display: flex;
  flex-direction: column;
}

.CurrentPlayer {
  margin: auto;
  position: absolute;
  bottom: 5vh;
  left: 20vw;
  right: 20vw;
  height: 10vh;
  width: 30vw;
  font-size: 3em;
  font-family: "Roboto", sans-serif;
  color: white;
}
