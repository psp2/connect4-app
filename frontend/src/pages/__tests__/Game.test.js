import React from "react";
import Game from "../Game/Game";
import renderer from "react-test-renderer";

it("game page, renders properly", () => {
  const state = jest.fn(() => {setGameState([1, []])});
  const tree = renderer
    .create(<Game/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});


it("game page, takes normal names 1", () => {
  const state = jest.fn(() => {setGameState([1, []])});
  let name1 = "Prashant";
  let name2 = "Vivek";
  const tree = renderer
    .create(<Game name1={name1} name2={name2}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});


it("game page, takes normal names 2", () => {
  const state = jest.fn(() => {setGameState([1, []])});
  let name1 = "Angrave";
  let name2 = "Chancellor Jones";
  const tree = renderer
    .create(<Game name1={name1} name2={name2}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("game page, null names", () => {
  const state = jest.fn(() => {setGameState([1, []])});
  const tree = renderer
    .create(<Game name1={null} name2={null}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
