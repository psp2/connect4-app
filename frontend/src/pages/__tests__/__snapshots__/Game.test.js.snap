// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game page, renders properly 1`] = `
<div
  className="Game"
>
  <div
    className="GameZone"
  >
    <div
      className="Player1_Zone"
    >
      <div
        className="PlayerIcon"
      >
        <div
          className="PlayerInfo"
        >
          <div>
            Player 
            1
            :
          </div>
          <div />
        </div>
        <div
          className="PlayerPicture1"
        />
        <div
          className="PlayerButtons"
        >
          <button
            className="Player1IconButton btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            UNDO
          </button>
          <button
            className="Player1IconButton btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            RESTART
          </button>
          <button
            className="Player1IconButton btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            FORFEIT
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        className="Board"
      />
      <div
        className="GameStartbutton"
        onClick={[Function]}
      >
        START
      </div>
    </div>
    <div
      className="Player2_Zone"
    >
      <div
        className="PlayerIcon"
      >
        <div
          className="PlayerInfo"
        >
          <div>
            Player 
            2
            :
          </div>
          <div />
        </div>
        <div
          className="PlayerPicture2"
        />
        <div
          className="PlayerButtons"
        >
          <button
            className="Player2IconButton btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            UNDO
          </button>
          <button
            className="Player2IconButton btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            RESTART
          </button>
          <button
            className="Player2IconButton btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            FORFEIT
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="CurrentPlayer"
  >
    Player Turn: 
    1
  </div>
</div>
`;
